File: packages/cli/src/services/skills-json.ts
import { readFile, writeFile, mkdir } from 'node:fs/promises';
import * as path from 'node:path';
import * as os from 'node:os';
import { SkillsJsonSchema, SkillsLockSchema } from '@spm/shared';
import type { SkillsJson, SkillsLock } from '@spm/shared';
import type { ResolvedSkill } from './resolver.js';

const SKILLS_JSON = 'skills.json';
const SKILLS_LOCK = 'skills-lock.json';

/**
 * Get the path for global skills.json (~/.spm/skills.json).
 */
export const getGlobalSkillsDir = (): string => {
  return path.join(os.homedir(), '.spm');
};

/**
 * Read and validate skills.json from a directory.
 */
export const loadSkillsJson = async (dir: string): Promise<SkillsJson | null> => {
  const filePath = path.join(dir, SKILLS_JSON);
  try {
    const raw = await readFile(filePath, 'utf-8');
    const parsed: unknown = JSON.parse(raw);
    const result = SkillsJsonSchema.safeParse(parsed);
    if (!result.success) {
      throw new Error(`Invalid ${SKILLS_JSON}: ${result.error.message}`);
    }
    return result.data;
  } catch (err) {
    if (err instanceof Error && 'code' in err && (err as NodeJS.ErrnoException).code === 'ENOENT') {
      return null;
    }
    throw err;
  }
};

/**
 * Write skills.json to a directory.
 */
export const saveSkillsJson = async (dir: string, data: SkillsJson): Promise<void> => {
  await mkdir(dir, { recursive: true });
  const filePath = path.join(dir, SKILLS_JSON);
  await writeFile(filePath, JSON.stringify(data, null, 2) + '\n', 'utf-8');
};

/**
 * Add a skill dependency to skills.json.
 * Creates the file if it doesn't exist.
 */
export const addSkillToJson = async (dir: string, name: string, range: string): Promise<void> => {
  let data = await loadSkillsJson(dir);
  if (!data) {
    data = { skills: {} };
  }
  data.skills[name] = range;
  await saveSkillsJson(dir, data);
};

/**
 * Remove a skill dependency from skills.json.
 */
export const removeSkillFromJson = async (dir: string, name: string): Promise<boolean> => {
  const data = await loadSkillsJson(dir);
  if (!data || !(name in data.skills)) {
    return false;
  }
  delete data.skills[name];
  await saveSkillsJson(dir, data);
  return true;
};

/**
 * Read and validate skills-lock.json from a directory.
 */
export const loadLockFile = async (dir: string): Promise<SkillsLock | null> => {
  const filePath = path.join(dir, SKILLS_LOCK);
  try {
    const raw = await readFile(filePath, 'utf-8');
    const parsed: unknown = JSON.parse(raw);
    const result = SkillsLockSchema.safeParse(parsed);
    if (!result.success) {
      throw new Error(`Invalid ${SKILLS_LOCK}: ${result.error.message}`);
    }
    return result.data;
  } catch (err) {
    if (err instanceof Error && 'code' in err && (err as NodeJS.ErrnoException).code === 'ENOENT') {
      return null;
    }
    throw err;
  }
};

/**
 * Write skills-lock.json to a directory.
 */
export const saveLockFile = async (dir: string, data: SkillsLock): Promise<void> => {
  await mkdir(dir, { recursive: true });
  const filePath = path.join(dir, SKILLS_LOCK);
  await writeFile(filePath, JSON.stringify(data, null, 2) + '\n', 'utf-8');
};

/**
 * Update lock file entries for resolved skills.
 * Merges new entries into existing lock file, or creates a new one.
 */
export const updateLockFile = async (dir: string, resolved: ResolvedSkill[]): Promise<void> => {
  let lock = await loadLockFile(dir);
  if (!lock) {
    lock = {
      lockfileVersion: 1,
      generated_at: new Date().toISOString(),
      generated_by: 'spm@0.0.1',
      skills: {},
    };
  }

  for (const skill of resolved) {
    lock.skills[skill.name] = {
      version: skill.version,
      resolved: skill.downloadUrl,
      checksum: skill.checksum,
      source: 'registry',
      signer: skill.signed ? undefined : undefined,
    };
  }

  lock.generated_at = new Date().toISOString();
  await saveLockFile(dir, lock);
};

/**
 * Remove a skill entry from the lock file.
 */
export const removeFromLockFile = async (dir: string, name: string): Promise<boolean> => {
  const lock = await loadLockFile(dir);
  if (!lock || !(name in lock.skills)) {
    return false;
  }
  delete lock.skills[name];
  lock.generated_at = new Date().toISOString();
  await saveLockFile(dir, lock);
  return true;
};

File: cli-go/internal/skillsjson/skillsjson.go
// Package skillsjson manages skills.json and skills-lock.json files.
// skills.json tracks installed skill dependencies (name -> version range).
// skills-lock.json locks resolved versions with checksums and metadata.
package skillsjson

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

const (
	SkillsJsonFile = "skills.json"
	SkillsLockFile = "skills-lock.json"
	LockVersion    = 1
)

// GeneratedBy is set at startup from cmd.Version via SetVersion.
var GeneratedBy = "spm@dev"

// SetVersion updates the GeneratedBy string used in lock files.
func SetVersion(v string) {
	GeneratedBy = "spm@" + v
}

// SkillsJson represents the skills.json file structure.
type SkillsJson struct {
	Skills map[string]string `json:"skills"`
}

// LockEntry represents a single locked skill entry.
type LockEntry struct {
	Version  string `json:"version"`
	Resolved string `json:"resolved"`
	Checksum string `json:"checksum"`
	Source   string `json:"source"`
	Signer   string `json:"signer,omitempty"`
}

// SkillsLock represents the skills-lock.json file structure.
type SkillsLock struct {
	LockfileVersion int                  `json:"lockfileVersion"`
	GeneratedAt     string               `json:"generated_at"`
	GeneratedBy     string               `json:"generated_by"`
	Skills          map[string]LockEntry `json:"skills"`
}

// ResolvedSkill holds data for a skill that has been resolved by the version resolver.
type ResolvedSkill struct {
	Name        string
	Version     string
	DownloadURL string
	Checksum    string
	Signed      bool
	Signer      string
}

// LoadSkillsJson reads and parses skills.json from the given directory.
// Returns nil, nil if the file does not exist.
func LoadSkillsJson(dir string) (*SkillsJson, error) {
	filePath := filepath.Join(dir, SkillsJsonFile)
	data, err := os.ReadFile(filePath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, fmt.Errorf("reading %s: %w", SkillsJsonFile, err)
	}

	var sj SkillsJson
	if err := json.Unmarshal(data, &sj); err != nil {
		return nil, fmt.Errorf("parsing %s: %w", SkillsJsonFile, err)
	}

	if sj.Skills == nil {
		sj.Skills = make(map[string]string)
	}

	return &sj, nil
}

// SaveSkillsJson writes skills.json to the given directory.
// Creates the directory if it does not exist.
func SaveSkillsJson(dir string, sj *SkillsJson) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("creating directory %s: %w", dir, err)
	}

	data, err := json.MarshalIndent(sj, "", "  ")
	if err != nil {
		return fmt.Errorf("marshaling %s: %w", SkillsJsonFile, err)
	}

	filePath := filepath.Join(dir, SkillsJsonFile)
	if err := os.WriteFile(filePath, append(data, '\n'), 0o644); err != nil {
		return fmt.Errorf("writing %s: %w", SkillsJsonFile, err)
	}

	return nil
}

// AddSkill adds or updates a skill entry in skills.json.
// Creates the file if it does not exist.
func AddSkill(dir string, name string, versionRange string) error {
	sj, err := LoadSkillsJson(dir)
	if err != nil {
		return err
	}
	if sj == nil {
		sj = &SkillsJson{Skills: make(map[string]string)}
	}

	sj.Skills[name] = versionRange
	return SaveSkillsJson(dir, sj)
}

// RemoveSkill removes a skill entry from skills.json.
// Returns false if the skill was not found.
func RemoveSkill(dir string, name string) (bool, error) {
	sj, err := LoadSkillsJson(dir)
	if err != nil {
		return false, err
	}
	if sj == nil {
		return false, nil
	}

	if _, exists := sj.Skills[name]; !exists {
		return false, nil
	}

	delete(sj.Skills, name)
	return true, SaveSkillsJson(dir, sj)
}

// LoadLockFile reads and parses skills-lock.json from the given directory.
// Returns nil, nil if the file does not exist.
func LoadLockFile(dir string) (*SkillsLock, error) {
	filePath := filepath.Join(dir, SkillsLockFile)
	data, err := os.ReadFile(filePath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) {
			return nil, nil
		}
		return nil, fmt.Errorf("reading %s: %w", SkillsLockFile, err)
	}

	var lock SkillsLock
	if err := json.Unmarshal(data, &lock); err != nil {
		return nil, fmt.Errorf("parsing %s: %w", SkillsLockFile, err)
	}

	if lock.Skills == nil {
		lock.Skills = make(map[string]LockEntry)
	}

	return &lock, nil
}

// SaveLockFile writes skills-lock.json to the given directory.
// Creates the directory if it does not exist.
func SaveLockFile(dir string, lock *SkillsLock) error {
	if err := os.MkdirAll(dir, 0o755); err != nil {
		return fmt.Errorf("creating directory %s: %w", dir, err)
	}

	data, err := json.MarshalIndent(lock, "", "  ")
	if err != nil {
		return fmt.Errorf("marshaling %s: %w", SkillsLockFile, err)
	}

	filePath := filepath.Join(dir, SkillsLockFile)
	if err := os.WriteFile(filePath, append(data, '\n'), 0o644); err != nil {
		return fmt.Errorf("writing %s: %w", SkillsLockFile, err)
	}

	return nil
}

// UpdateLockFile merges resolved skills into the lock file.
// Preserves existing entries and adds/updates changed ones.
// Creates the file if it does not exist.
func UpdateLockFile(dir string, resolved []ResolvedSkill) error {
	lock, err := LoadLockFile(dir)
	if err != nil {
		return err
	}
	if lock == nil {
		lock = &SkillsLock{
			LockfileVersion: LockVersion,
			Skills:          make(map[string]LockEntry),
		}
	}
	// Always update metadata on write
	lock.GeneratedAt = time.Now().UTC().Format(time.RFC3339)
	lock.GeneratedBy = GeneratedBy

	for _, skill := range resolved {
		entry := LockEntry{
			Version:  skill.Version,
			Resolved: skill.DownloadURL,
			Checksum: skill.Checksum,
			Source:   "registry",
		}
		if skill.Signer != "" {
			entry.Signer = skill.Signer
		}
		lock.Skills[skill.Name] = entry
	}

	lock.GeneratedAt = time.Now().UTC().Format(time.RFC3339)
	return SaveLockFile(dir, lock)
}

// RemoveFromLockFile removes a skill entry from the lock file.
// Returns false if the skill was not found.
func RemoveFromLockFile(dir string, name string) (bool, error) {
	lock, err := LoadLockFile(dir)
	if err != nil {
		return false, err
	}
	if lock == nil {
		return false, nil
	}

	if _, exists := lock.Skills[name]; !exists {
		return false, nil
	}

	delete(lock.Skills, name)
	lock.GeneratedAt = time.Now().UTC().Format(time.RFC3339)
	return true, SaveLockFile(dir, lock)
}
